File: beware-of-the-sheep/src/Entities/Plants/Hogweed.cpp
#include "pch.h"
#include "Hogweed.h"
#include "Entities/Animals/CyberSheep.h"
#include "World.h"

namespace bots {

	Hogweed::Hogweed(World & world, const Point & position) :
		Plant(world, position)
	{}

	void Hogweed::action()
	{
		long lxpos = static_cast<long>(getPosition().x), lypos = static_cast<long>(getPosition().y);
		unsigned int xpos = getPosition().x, ypos = getPosition().y;
		unsigned int width = m_world.getWidth(), height = m_world.getHeight();

		unsigned int nx = (lxpos - 1 < 0 ? lxpos - 1 + width : (lxpos - 1) % width);
		auto e = m_world.getEntityAt(nx, ypos);
		if (dynamic_cast<Animal*>(e.get()) && e->getKind() != Entity::Kind::Hogweed && e->getKind() != Entity::Kind::CyberSheep)
			e->kill();

		nx = (lxpos + 1 < 0 ? lxpos + 1 + width : (lxpos + 1) % width);
		e = m_world.getEntityAt(nx, ypos);
		if (dynamic_cast<Animal*>(e.get()) && e->getKind() != Entity::Kind::Hogweed && e->getKind() != Entity::Kind::CyberSheep)
			e->kill();

		unsigned int ny = (lypos - 1 < 0 ? lypos - 1 + height : (lypos - 1) % height);
		e = m_world.getEntityAt(xpos, ny);
		if (dynamic_cast<Animal*>(e.get()) && e->getKind() != Entity::Kind::Hogweed && e->getKind() != Entity::Kind::CyberSheep)
			e->kill();

		ny = (lypos + 1 < 0 ? lypos + 1 + height : (lypos + 1) % height);
		e = m_world.getEntityAt(xpos, ny);
		if (dynamic_cast<Animal*>(e.get()) && e->getKind() != Entity::Kind::Hogweed && e->getKind() != Entity::Kind::CyberSheep)
			e->kill();

		Plant::action();
	}

	void Hogweed::collision(Entity & other)
	{
		if (!dynamic_cast<CyberSheep *>(&other))
			other.kill();
	}

} // namespace bots

File: src/com/kostu96/bots/entities/plants/Hogweed.java
package com.kostu96.bots.entities.plants;

import com.kostu96.bots.World;
import com.kostu96.bots.entities.Entity;
import com.kostu96.bots.entities.animals.Animal;
import com.kostu96.bots.entities.animals.CyberSheep;
import com.kostu96.bots.utils.ImageManager.ImageID;
import com.kostu96.bots.utils.Point;

public class Hogweed extends Plant {
	private static final int STRENGTH = 10;

	public Hogweed(World world, Point position) {
		super(world, position, STRENGTH);
	}

	@Override
	public void action() {
		super.action();
		
		int xpos = getPosition().x, ypos = getPosition().y;
		int width = m_world.getColumns(), height = m_world.getRows();

		int nx = (xpos - 1 < 0 ? xpos - 1 + width : (xpos - 1) % width);
		Entity e = m_world.getEntityAt(nx, ypos);
		if (e instanceof Animal && !(e instanceof Hogweed) && !(e instanceof CyberSheep))
			e.kill();

		nx = (xpos + 1 < 0 ? xpos + 1 + width : (xpos + 1) % width);
		e = m_world.getEntityAt(nx, ypos);
		if (e instanceof Animal && !(e instanceof Hogweed) && !(e instanceof CyberSheep))
			e.kill();

		int ny = (ypos - 1 < 0 ? ypos - 1 + height : (ypos - 1) % height);
		e = m_world.getEntityAt(xpos, ny);
		if (e instanceof Animal && !(e instanceof Hogweed) && !(e instanceof CyberSheep))
			e.kill();

		ny = (ypos + 1 < 0 ? ypos + 1 + height : (ypos + 1) % height);
		e = m_world.getEntityAt(xpos, ny);
		if (e instanceof Animal && !(e instanceof Hogweed) && !(e instanceof CyberSheep))
			e.kill();
	}
	
	@Override
	public void collision(Entity other) {
		if (!(other instanceof CyberSheep))
			other.kill();
	}

	@Override
	public ImageID getImageID() {
		return ImageID.HOGWEED;
	}

}
